[temp.func.order] Specify to only add extra first argument if needed CWG2834

HoBoIs opened the pull request:
In [temp.func.order] paragraph 3, it is written that we need to add a new first parameter for all member functions for the purposes of template deduction. However, it only makes sense if we have an implicit object parameter.

If we don't have an implicit object parameter, we shouldn't add a new first parameter because all of our parameters explicitly are written.

HoBoIs commented:
@jwiegley could you take a look, or recommend someone who can take a look?

jwiegley commented:
I haven't been involved with the C++ standard for 10 years now, I do not have any current recommendations.

jensmaurer commented:
Please squash the commits and force-push.

This looks like a CWG defect to me.

HoBoIs committed:
[temp.func.order] Explicitly specify to only add extra first argument if we don't have an explicit object argument

In [temp.func.order] paragraph 3, it is written that we need to add a new first parameter for all member functions for the purposes of template deduction. However, it only makes sense if we don't have an explicit object parameter.

If we have an explicit object parameter, we shouldn't add a new (implicit) first parameter because our first parameter is explicitly written.

HoBoIs commented:
@jensmaurer
I have squashed and force-pushed them.

HoBoIs commented:
[This](https://www.open-std.org/jtc1/sc22/wg21/docs/cwg_active.html#2834) open cwg defect solves the same problem, by suggesting the same change (and some more for static member functions). So I think my change does not need to be a separate thing on its own.

jensmaurer commented:
CWG2834
